### DISCUSSION TITLE
is It possible to programmatically capture Saml Assertion from external Idp?

### DISCUSSION THREAD

**Discussion by PavelJacobo** [aws/aws-aspnet-cognito-identity-provider#232] [Q&A]
Hello, I'm using Cognito and Okta as external SAML Idp. So when user wants to access my application it's redirected to Okta that is acting as Idp.
Okta sends a SAML response with SAML assertion back to Cognito to /idpresponse endpoint. I can see SAML response in the browser, but I'm not able to capture this SamlResponse **programatically** to be able to retrieve a token from Okta Oauth server.

I want to implement the following flow [here](https://developer.okta.com/docs/guides/implement-grant-type/saml2assert/main/#grant-type-flow):

Saml assertion flow with grant-type: urn:ietf:params:oauth:grant-type:saml2-bearer

this implies that we have to be able to get SAML assertion and send it back to Okta OAuth to /token endpoint.

How can we programmatically capture SAML response from Idp in Cognito to be able to implement SAML assertion grant flow as described in Okta documentation? I've tried with Post authentication Lambda trigger and Pre authentication Lambda trigger, but SamlReponse parameters do not end on any of the lambdas. 

this is the flow:
![image](https://github.com/aws/aws-aspnet-cognito-identity-provider/assets/38545024/6deb6f70-b335-4173-8fa5-de987f261eff)

this is the flow I would like to implement. I need to capture SAML assertion, because I also need a token from external Okta OAuth.
![image](https://github.com/aws/aws-aspnet-cognito-identity-provider/assets/38545024/7010ff1c-1919-410a-a8c9-76383067f5b4)

Thanks so much

**Comment by ashishdhingra** [comment]
@PavelJacobo Good morning. I'm unsure if this could help, you may try [Customizing user pool workflows with Lambda triggers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html).

**Reply by PavelJacobo** [reply]
Hello, Thanks for answer

I've tried with Post authentication Lambda trigger and Pre authentication Lambda trigger, but SamlReponse parameters do not end on any of the lambdas.

I cannot find no info in the documentation related to how to capture the outboud request to /saml2/idpresponse

for example this is the **event** I receive in a **Post authentication Lambda trigger** Lambda
as you can see, the event has no information about the http request parameters
```
{
    "version": "1",
    "region": "us-east-1",
    "userPoolId": "us-east-1_xyzxyz",
    "userName": "oktassopoc_johndoe@xxx.com",
    "callerContext": {
        "awsSdkVersion": "aws-sdk-unknown-unknown",
        "clientId": "69pf1gxyzrk9obqxyz"
    },
    "triggerSource": "PostAuthentication_Authentication",
    "request": {
        "userAttributes": {
            "sub": "6zxy67f-9011-4999-xyz-52xyzf51",
            "custom:Idp_Id": "00u8xxyz123gdOw5d7",
            "email_verified": "false",
            "identities": "[{\"userId\":\"pgonzalez@xxx.com\",\"providerName\":\"oktassopoc\",\"providerType\":\"SAML\",\"issuer\":\"http://www.okta.com/xyz\",\"primary\":true,\"dateCreated\":1680689741221}]",
            "cognito:user_status": "EXTERNAL_PROVIDER",
            "name": "Pavel xxx",
            "middle_name": "xxx",
            "given_name": "Pavel",
            "email": "jonhdoe@xxx.com"
        },
        "newDeviceUsed": false
    },
    "response": {}
}
```

Thanks so much

regards

**Reply by ashishdhingra** [reply]
@PavelJacobo Good afternoon. I was going though open discussions and came across this one. One of the solution could be to `Using Amazon Cognito Identity Pools with Direct SAML Federation`. Instead of using Cognito User Pools as your SAML Service Provider, you'll need to:

- Configure Cognito Identity Pools to federate directly with Okta via SAML
- Implement custom authentication logic that captures the SAML assertion
- Exchange the assertion with Okta's OAuth endpoint before passing credentials to Cognito Identity Pools
